### PR TITLE
CO-152 prevent duplicate control-host intake owners

### PR DESCRIPTION
**Pull request by Kbediako**
## WHAT?
- Add a same-task control-host ownership lock and local owner diagnostics for provider intake.
- Expose owner metadata through provider polling/status compatibility surfaces and provider-worker refresh timeout warnings.
- Add the CO-152 docs-first packet plus focused ownership, lifecycle, runtime, and provider-worker regressions.

## WHY?
- A duplicate `control-host --task local-mcp --run control-host --pipeline provider-linear-worker` process can otherwise compete on the same provider-intake state.
- Duplicate provider refreshers can waste Linear/GitHub budget and make stuck lifecycle diagnostics ambiguous.
- The fix needs to reject or reclaim ownership without killing provider workers or mutating provider-intake claims.

## HOW?
- Acquire an atomic `control-host-owner.lock` before startup input preparation, ready-instance startup, and provider refresh scheduling.
- Reject active or ambiguous owners with `control-host-duplicate-owner.json`; reclaim stale owners only after PID liveness proves inactivity.
- Release only the current owner token, write the public released owner artifact before removing the lock, and clear stale diagnostics on later clean acquisition.
- Persist additive `control_host_owner` polling metadata and append ownership hints to provider-worker refresh-request failures.

Diff budget override: CO-152 requires docs-first packet plus new ownership helper/tests across startup/status/provider-worker warning surfaces; split would leave duplicate-host guard unverifiable.

## BEFORE
```ts
const startupInputs = await prepareControlServerStartupInputs(...);
const readyInstance = await startControlServerReadyInstanceLifecycle(...);
const providerRefreshStartupTrigger = scheduleStartupProviderRefresh(...);
```

## AFTER
```ts
const owner = await acquireControlHostOwnership(...);
const startupInputs = await prepareControlServerStartupInputs(...);
const readyInstance = await startControlServerReadyInstanceLifecycle(...);
```

## TESTS
- [x] `npx vitest run --config vitest.config.core.ts orchestrator/tests/ControlHostOwnership.test.ts orchestrator/tests/ControlServerPublicLifecycle.test.ts orchestrator/tests/ControlRuntime.test.ts orchestrator/tests/ProviderLinearWorkerRunner.test.ts`
- [x] `node scripts/delegation-guard.mjs`
- [x] `node scripts/spec-guard.mjs --dry-run`
- [x] `npm run build`
- [x] `npm run lint`
- [x] `npm run test`
- [x] `npm run docs:check`
- [x] `npm run docs:freshness`
- [x] `npm run repo:stewardship`
- [x] `DIFF_BUDGET_OVERRIDE_REASON="CO-152 requires docs-first packet plus new ownership helper/tests across startup/status/provider-worker warning surfaces; split would leave duplicate-host guard unverifiable." node scripts/diff-budget.mjs`
- [x] `TASK=linear-0d66d189-fc51-4054-80db-b6990858f33d ... FORCE_CODEX_REVIEW=1 ... npm run review -- --task linear-0d66d189-fc51-4054-80db-b6990858f33d --uncommitted` executed and failed closed at `review_outcome: failed-boundary` / `termination_boundary.kind: command-intent`; manual correctness/regressions/missing-tests review completed and fixed the release-order artifact risk.
- [x] Explicit elegance/minimality pass completed after manual review.
- [x] `npm run pack:smoke`
- [x] `git diff --check`

## FIXES
Fixes https://linear.app/asabeko/issue/CO-152/co-prevent-duplicate-same-task-control-hosts-from-competing-on

## Summary by CodeRabbit

* **New Features**
  * Exclusive control‑host ownership lock with fail‑closed duplicate detection, deterministic stale‑owner reclamation and operator‑visible diagnostics.

* **Documentation**
  * Added PRD, tech spec, action plan, task checklists and freshness registry entries detailing requirements, rollout, validation, risks and handoff.

* **Behaviour/Runtime**
  * Ownership acquired before provider refresh, surfaced via polling/state, and safely released on shutdown with guarded error handling.

* **Tests**
  * New tests covering acquisition, duplicate rejection, stale reclamation, diagnostics and lifecycle interactions.

### CODE REVIEW

File: orchestrator/src/cli/control/controlServerPublicLifecycle.ts [hunk 1]
```diff
@@ -191,11 +220,26 @@ export async function closeControlServerPublicLifecycle(
...
+  try {
+    await state.controlHostOwnership?.release();
+  } catch (error) {
+    if (!closeError) {
+      throw error;
+    }
+  }
+  if (closeError) {
+    throw closeError;
+  }
```
[the remark this answers is not in the document]
Comment (Kbediako) [reply orchestrator/src/cli/control/controlServerPublicLifecycle.ts:235]: @coderabbitai Fixed in ed8f6b59170c071c264d170b6809ab6a215ca429: close now keeps the ownership lock in place when runtime shutdown fails, and only releases after runtime close succeeds. Added a regression test for the failed-shutdown path. Please re-review/resolve.

File: tasks/tasks-linear-0d66d189-fc51-4054-80db-b6990858f33d.md [hunk 2]
```diff
@@ -0,0 +1,74 @@
...
+
+## Handoff
+- [ ] PR attached to the issue. Evidence: pending.
+- [ ] Latest `origin/main` merged into the branch before review-state transition. Evidence: pending.
+- [ ] PR checks green and `pr ready-review` drain clean before review-state transition. Evidence: pending.
+- [ ] Unresolved actionable review threads: `0` or explicit pushback recorded. Evidence: pending.
+- [ ] Issue moved to `In Review`. Evidence: pending.
+
+## Progress Log
+- 2026-04-11: Issue moved to `In Progress`, parallelization decision recorded as serial due overlapping owner/status/provider-intake contract, workpad created, and branch created from `origin/main`.
```
[the remark this answers is not in the document]
Comment (Kbediako) [reply tasks/tasks-linear-0d66d189-fc51-4054-80db-b6990858f33d.md]: @coderabbitai Fixed in ed8f6b59170c071c264d170b6809ab6a215ca429: corrected the checklist wording from `due overlapping` to `due to overlapping` in both task mirrors. Please re-review/resolve.

File: orchestrator/src/cli/control/controlHostOwnership.ts [hunk 3]
```diff
@@ -0,0 +1,665 @@
...
+        await writeControlHostOwnershipDiagnostic({
+          paths,
+          reason: 'stale_control_host_owner',
+          action: 'stale_reclaimed',
+          existingOwner: existingState.owner,
+          attemptedOwner,
+          observedAt: now(),
+          diagnosticPath: paths.staleDiagnosticPath
+        });
+        await rm(paths.lockDir, { recursive: true, force: true });
```
[the remark this answers is not in the document]
Comment (Kbediako) [reply orchestrator/src/cli/control/controlHostOwnership.ts]: Fixed in c9282778f2d82f56a3ccc0e116396ce571dcacb4: stale reclaim now uses a short-lived reclaim lock, rechecks that the same stale owner token is still present under that guard, and retries/fails closed instead of deleting if the owner changed or another reclaimer is active. Added a regression test that swaps in a replacement owner after the stale read and verifies the replacement lock survives.

File: tasks/tasks-linear-0d66d189-fc51-4054-80db-b6990858f33d.md [hunk 4]
```diff
@@ -0,0 +1,74 @@
...
+- [x] `npm run lint`. Evidence: passed after the manual-review patch.
+- [x] `npm run test`. Evidence: passed 329 files / 3501 tests after the manual-review patch.
+- [x] `npm run docs:check`. Evidence: passed with `docs:check: OK`.
+- [x] `npm run docs:freshness`. Evidence: passed with 3578 docs / 3581 registry entries.
+- [x] `npm run repo:stewardship`. Evidence: passed with 4550 tracked files and existing 7 action-required archive/reference items.
+- [x] `node scripts/diff-budget.mjs`. Evidence: passed with `DIFF_BUDGET_OVERRIDE_REASON` because CO-152 needs docs-first packet plus ownership helper/tests across startup/status/provider-worker warning surfaces; split would leave the duplicate-host guard unverifiable.
+- [x] Manifest-backed standalone review plus explicit elegance review before review handoff. Evidence: wrapper executed under `FORCE_CODEX_REVIEW=1` and failed closed as `review_outcome: failed-boundary` / `termination_boundary.kind: command-intent`; manual correctness/regressions/missing-tests review found and fixed a release-order artifact overwrite risk, then explicit elegance review kept the focused helper and removed no further scope.
```
[the remark this answers is not in the document]
Comment (Kbediako) [reply tasks/tasks-linear-0d66d189-fc51-4054-80db-b6990858f33d.md]: @coderabbitai Fixed in 6207f3c0a43cfe30689280af615e2e4bc7d79a25: the task and .agent mirrors now record this as a standalone-review fallback waiver with the failed-boundary evidence as the reason, instead of claiming the manifest-backed wrapper completed successfully. The focused test count was also refreshed to 258 after the later feedback fixes. Please re-review/resolve.

File: .agent/task/linear-0d66d189-fc51-4054-80db-b6990858f33d.md [hunk 5]
```diff
@@ -0,0 +1,75 @@
...
+- [x] `npm run docs:check`. Evidence: passed with `docs:check: OK`.
+- [x] `npm run docs:freshness`. Evidence: passed with 3578 docs / 3581 registry entries.
+- [x] `npm run repo:stewardship`. Evidence: passed with 4550 tracked files and existing 7 action-required archive/reference items.
+- [x] `node scripts/diff-budget.mjs`. Evidence: passed with `DIFF_BUDGET_OVERRIDE_REASON` because CO-152 needs docs-first packet plus ownership helper/tests across startup/status/provider-worker warning surfaces; split would leave the duplicate-host guard unverifiable.
+- [x] Standalone review fallback waiver plus explicit elegance review before review handoff. Evidence: wrapper executed under `FORCE_CODEX_REVIEW=1` and failed closed as `review_outcome: failed-boundary` / `termination_boundary.kind: command-intent`; waiver recorded because the wrapper reached no verdict, then manual correctness/regressions/missing-tests review found and fixed a release-order artifact overwrite risk and explicit elegance review kept the focused helper without adding broader process-kill, claim-mutation, or budget/backoff scope.
```
[the remark this answers is not in the document]
Comment (Kbediako) [reply .agent/task/linear-0d66d189-fc51-4054-80db-b6990858f33d.md]: @coderabbitai Fixed in 05d6b62e505c6c48457729290ed2adb27e2f5f23: the .agent mirror now leads with the waiver decision and rationale, with failed-boundary telemetry moved to supporting context. Please re-review/resolve.

File: tasks/tasks-linear-0d66d189-fc51-4054-80db-b6990858f33d.md [hunk 6]
```diff
@@ -0,0 +1,75 @@
...
+- [x] `npm run docs:check`. Evidence: passed with `docs:check: OK`.
+- [x] `npm run docs:freshness`. Evidence: passed with 3578 docs / 3581 registry entries.
+- [x] `npm run repo:stewardship`. Evidence: passed with 4550 tracked files and existing 7 action-required archive/reference items.
+- [x] `node scripts/diff-budget.mjs`. Evidence: passed with `DIFF_BUDGET_OVERRIDE_REASON` because CO-152 needs docs-first packet plus ownership helper/tests across startup/status/provider-worker warning surfaces; split would leave the duplicate-host guard unverifiable.
+- [x] Standalone review fallback waiver plus explicit elegance review before review handoff. Evidence: wrapper executed under `FORCE_CODEX_REVIEW=1` and failed closed as `review_outcome: failed-boundary` / `termination_boundary.kind: command-intent`; waiver recorded because the wrapper reached no verdict, then manual correctness/regressions/missing-tests review found and fixed a release-order artifact overwrite risk and explicit elegance review kept the focused helper without adding broader process-kill, claim-mutation, or budget/backoff scope.
```
[the remark this answers is not in the document]
Comment (Kbediako) [reply tasks/tasks-linear-0d66d189-fc51-4054-80db-b6990858f33d.md]: @coderabbitai Fixed in 05d6b62e505c6c48457729290ed2adb27e2f5f23: the task mirror now leads with the waiver decision and rationale, with failed-boundary telemetry moved to supporting context. Please re-review/resolve.

**Review by Kbediako — COMMENTED** [review]
Submitting pending thread acknowledgements for addressed bot review feedback.

**Review by Kbediako — COMMENTED** [review]
Submitting pending thread acknowledgement for addressed bot review feedback.

File: orchestrator/src/cli/control/controlHostOwnership.ts [hunk 7]
```diff
@@ -0,0 +1,665 @@
...
+        continue;
```
[the remark this answers is not in the document]
Comment (Kbediako) [reply orchestrator/src/cli/control/controlHostOwnership.ts:196]: @coderabbitai Fixed in c9282778f2d82f56a3ccc0e116396ce571dcacb4: stale reclaim now uses a short-lived reclaim lock, rechecks that the same stale owner token is still present under that guard, and refuses to delete if the owner changed or another reclaimer is active. Latest head 05d6b62 includes the fix. Please re-check/resolve.
